### PR TITLE
DM-34578: Initial implementation of Scheduler predicted queue feature

### PR DESCRIPTION
**Pull request by tribeiro**

### CODE REVIEW

File: doc/version-history.rst [hunk 1]
```diff
@@ -4,6 +4,30 @@
...
+-------
+
+* In Scheduler CSC:
+
+  * Add new command `computePredictedSchedule`, using the new `support_command` utility to maintain backward compatibility.
+  * Publish general info.
+  * Publish time to next target.
+  * Add `compute_predicted_schedule` feature.
+    The method runs in the advance control loop just after generating the target queue.
+  * Add `_get_targets_in_time_window`, to run the scheduler into the future until it produces a set number of targets or reach the end of the specified time window.
```
Comment (wvreeven) [inline doc/version-history.rst:17]: `reach` -> `reaches`

File: python/lsst/ts/scheduler/utils/csc_utils.py [hunk 2]
```diff
@@ -102,3 +104,20 @@ def is_valid_efd_query(entry: str) -> bool:
...
+def support_command(command_name: str) -> bool:
+    """Check if CSC supports a particular command.
+
+    This is used to provide backward compatibility for new commands being added
+    to the CSC.
+
+    Returns
+    -------
+    `bool`
+        True if CSC interfaces defines the command "interruptWEP", False
```
Comment (wvreeven) [inline python/lsst/ts/scheduler/utils/csc_utils.py:118]: `CSC` -> `the CSC` and either `interfaces` -> `interface` or `defines` -> `define`

File: python/lsst/ts/scheduler/scheduler_csc.py [hunk 3]
```diff
@@ -208,6 +216,12 @@ def __init__(
         # The maximum tolerable time without targets in seconds.
         self.max_time_no_target = 3600.0
 
+        # The maximum number of targets in the predicted schedule.
+        self.max_predicted_targets = 1000
+
+        # Default command response timeout
```
Comment (wvreeven) [inline python/lsst/ts/scheduler/scheduler_csc.py:222]: Please end the sentence with `in seconds.`

File: python/lsst/ts/scheduler/scheduler_csc.py [hunk 4]
```diff
@@ -208,6 +216,12 @@ def __init__(
...
+        self.default_command_timeout = 60.0
+
         # How long to wait for target loop to stop before killing it
```
Comment (wvreeven) [inline python/lsst/ts/scheduler/scheduler_csc.py:225]: Not part of the code changes but please end this sentence with `in seconds.` as well.

File: python/lsst/ts/scheduler/scheduler_csc.py [hunk 5]
```diff
@@ -492,8 +533,42 @@ async def do_load(self, data):
...
+            timeout=self.default_command_timeout,
+            result="Loading snapshot.",
+        )
+
         await self._handle_load_snapshot(data.uri)
 
+    async def _do_computePredictedSchedule(self, data):
+        """Compute and publish the predicted schedule.
+
+        This command can only be executed if the Scheduler is idle. It it is
```
Comment (wvreeven) [inline python/lsst/ts/scheduler/scheduler_csc.py:547]: `It it` -> `If it`

File: python/lsst/ts/scheduler/scheduler_csc.py [hunk 6]
```diff
@@ -492,8 +533,42 @@ async def do_load(self, data):
...
+        data : `DataType`
+            Command data.
+        """
+
+        self.assert_enabled()
+
+        if self.run_target_loop.is_set():
+            raise RuntimeError(
+                "Scheduler is currently running. Note that the predicted schedule "
+                "is computed automatically while the Scheduler is running."
```
Comment (wvreeven) [inline python/lsst/ts/scheduler/scheduler_csc.py:561]: Perhaps I misunderstood but isn't the predicted schedule computed while the Scheduler is NOT running?

File: python/lsst/ts/scheduler/scheduler_csc.py [hunk 7]
```diff
@@ -1471,76 +1551,90 @@ async def generate_target_queue(self):
...
 
-            # Inside the loop we are running update_conditions directly from
-            # the driver instead if update_telemetry. This bypasses the updates
-            # done by the CSC method.
-            await loop.run_in_executor(None, self.driver.update_conditions)
+            while len(self.targets_queue) <= self.parameters.n_targets + 1:
 
-            target = await loop.run_in_executor(None, self.driver.select_next_target)
+                # Inside the loop we are running update_conditions directly
+                # from the driver instead if update_telemetry. This bypasses
```
Comment (wvreeven) [inline python/lsst/ts/scheduler/scheduler_csc.py:1596]: `if` -> `of`

File: python/lsst/ts/scheduler/scheduler_csc.py [hunk 8]
```diff
@@ -1560,10 +1654,44 @@ async def check_targets_queue_condition(self):
...
             self.targets_queue_condition.set_result(None)
 
-    async def save_scheduler_state(self):
-        """Save scheduler state to S3 bucket and publish event."""
+    async def save_scheduler_state(self, publish_lfoa):
+        """Save scheduler state to S3 bucket and publish event.
+
+        Parameters
+        ----------
+        publish_lfoa : bool
```
Comment (wvreeven) [inline python/lsst/ts/scheduler/scheduler_csc.py:1662]: Maybe surround `bool` with back ticks?

File: python/lsst/ts/scheduler/scheduler_csc.py [hunk 9]
```diff
@@ -1964,3 +2194,70 @@ async def _register_observations(
...
+            normally.
+        ra : float
+            Estimated RA of the target (in degrees).
+        dec : float
+            Estimated Declination of the target (in degrees).
+        rot_sky_pos : float
+            Estimated rotation angle (in degrees).
+        """
+
+        # TODO: Remove backward compatibility.
```
Comment (wvreeven) [inline python/lsst/ts/scheduler/scheduler_csc.py:2218]: TODO requires a Jira ticket.

**Review by wvreeven — APPROVED** [review]
Mostly typo comments though I'd strongly suggest to open a Jira ticket for the two TODOs and mention the Jira ticket number in the TODO comments. Otherwise looks good to me.

File: python/lsst/ts/scheduler/scheduler_csc.py (hunk 6 above)
Comment (tribeiro) [reply python/lsst/ts/scheduler/scheduler_csc.py:561]: No, when the Scheduler is running it computes the predicted queue and publishes the information. But this command allows one to request the Scheduler to compute and publish the information when it is NOT running. 

So, for instance, if we stopped the Scheduler for some engineering and we want to see what is coming up next, while we finish the engineering task, we can send this command.

**Pull request merged**